codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java

```diff
@@ -498,4 +478,51 @@ static long decodeULE128(ByteBuf in, long result) throws Http2Exception {
 
         throw DECODE_ULE_128_DECOMPRESSION_EXCEPTION;
     }
+
+    private static final class HeadersDecodingState {
+        private final Http2Headers headers;
+        private final int streamId;
+        private final long maxHeaderListSizeGoAway;
+        private final long maxHeaderListSize;
+
+        private boolean sawRegularHeader;
+        private long headersLength;
+
+        HeadersDecodingState(final int streamId, final Http2Headers headers,
+                                     final long maxHeaderListSizeGoAway, final long maxHeaderListSize) {
+            this.streamId = streamId;
+            this.headers = headers;
+            this.maxHeaderListSizeGoAway = maxHeaderListSizeGoAway;
+            this.maxHeaderListSize = maxHeaderListSize;
+        }
+
+        void addHeader(CharSequence name, CharSequence value) throws Http2Exception {
+            long newHeaderLength = this.headersLength + name.length() + value.length();
+            if (newHeaderLength > maxHeaderListSizeGoAway) {
+                headerListSizeExceeded(maxHeaderListSizeGoAway);
+            }
+
+            boolean pseudoHeaderFormat = hasPseudoHeaderFormat(name);
+            if (sawRegularHeader && pseudoHeaderFormat) {
+                throw connectionError(PROTOCOL_ERROR, "Pseudo-header field found after regular header.");
```

Comment:
It seems like we validate the pseudo header requirements in `DefaultHttp2HeadersDecoder` and in here in `HpackDecoder`. Can we validate in one or the other instead of both?

Reply:
Agree, I wanted to put everything into `DefaultHttp2HeadersDecoder` but the way `DefaultHttp2Headers` works we lose the insertion order. 

Do you think it would be better to have everything in `HpackDecoder` or should I find a way to retrieve insertion order from DefaultHttp2Headers?